### PR TITLE
Update to matrixSSL 3.8.4

### PR DESCRIPTION
**Pull request by vonydev**
Lots of changes to the XS module internal structures, how callbacks are handled.
Documentation update

### CODE REVIEW

File: MatrixSSL3.xs [hunk 1]
```diff
@@ -1309,15 +1447,14 @@ SV *get_master_secret(ssl)
     Crypt_MatrixSSL3_Sess *ssl;
 
     CODE:
-    RETVAL = newSVpv(matrixSslGetMasterSecret(ssl), SSL_HS_MASTER_SIZE);
+    RETVAL = NULL; //newSVpv(matrixSslGetMasterSecret(ssl), SSL_HS_MASTER_SIZE);
```
Comment (powerman) [inline MatrixSSL3.xs]: Are you sure this is correct change and not a debugging leftover?

File: lib/Crypt/MatrixSSL3.pm [hunk 2]
```diff
@@ -977,46 +1000,17 @@ More information about callback L</certValidator> in the L</CALLBACKS> section.
...
-
-    # 2) many times, after clients are accepted and server SSL sessions
-    #    created
-    $ssl->init_SNI( $sni_index, $ssl_id );
+    $ssl->init_SNI( $sserver_index, $sni_params );
 
-This will skip the SNI server initialization part and just register the
-MatrixSSL SNI callback to an internal XS function using the SNI server
-structure specified by $sni_index as parameter.
+This function shoule be called only once when the server is initialzied.
```
Comment (powerman) [inline lib/Crypt/MatrixSSL3.pm]: A couple of typos in this line.

File: lib/Crypt/MatrixSSL3.pm [hunk 3]
```diff
@@ -830,10 +833,17 @@ More information about L</VHIndexCallback> in the L</CALLBACKS> section.
...
 
 More information about L</ALPNCallback> in the L</CALLBACKS> section.
 
+=head2 create_SSL_server
+
+    $server_index = create_SSL_server();
+
+Tells the XS module to allocate a new server structure. The returned index
+must be saved and then used one time to initialize the server structure and then
+each time a nw client connection is accepted in order to set SNI/ALPN callbacks.
```
Comment (powerman) [inline lib/Crypt/MatrixSSL3.pm]: nw? Probably new?

**Comment by powerman** [comment]
Let me know when this PR will be ready to merge and release.

**Comment by powerman** [comment]
Also, it has conflicts with current master (latest released version) - please merge or rebase master first.

**Comment by vonydev** [comment]
All fixed. Let me know if you can pull the request now.

**Comment by powerman** [comment]
Thanks, I'll check it later today or tomorrow.

**Comment by powerman** [comment]
Released v3.8.0. Thanks @vonydev!

**Pull request merged**